arrow-test/src/main/kotlin/arrow/test/generators/GenK.kt

```diff
@@ -130,3 +142,46 @@ fun Try.Companion.genK() = object : GenK<ForTry> {
         Success(it)
       }, Gen.throwable().map { Try.Failure(it) })
 }
+
+// FIXME This is a bad gen and should be replaced by a proper function generator as soon as we use arrow-check in places
+fun <F, D> Kleisli.Companion.genK(genkF: GenK<F>) = object : GenK<KleisliPartialOf<F, D>> {
+  override fun <A> genK(gen: Gen<A>): Gen<Kind<KleisliPartialOf<F, D>, A>> = genkF.genK(gen).map { k ->
+    Kleisli { _: D -> k }
+  }
+}
+
+fun <F> OptionT.Companion.genK(genKF: GenK<F>): GenK<OptionTPartialOf<F>> = object : GenK<OptionTPartialOf<F>> {
+  override fun <A> genK(gen: Gen<A>): Gen<Kind<OptionTPartialOf<F>, A>> =
+    genKF.genK(Gen.option(gen)).map(::OptionT)
+}
+
+// FIXME This is a bad generator and should be replaced by a proper function generator as soon as we use arrow-check
+fun <F, S> StateT.Companion.genK(genkF: GenK<F>, genS: Gen<S>) = object : GenK<StateTPartialOf<F, S>> {
+  override fun <A> genK(gen: Gen<A>): Gen<Kind<StateTPartialOf<F, S>, A>> =
+    genkF.genK(Gen.tuple2(genS, gen)).map { state ->
+      StateT { _: S -> state }
+    }
+}
+
+fun <F, W> WriterT.Companion.genK(
+  GENKF: GenK<F>,
+  GENW: Gen<W>
+): GenK<WriterTPartialOf<F, W>> = object : GenK<WriterTPartialOf<F, W>> {
+  override fun <A> genK(gen: Gen<A>): Gen<Kind<WriterTPartialOf<F, W>, A>> =
+    GENKF.genK(Gen.tuple2(GENW, gen)).map(::WriterT)
+}
+
+fun IO.Companion.genK() = object : GenK<ForIO> {
+  override fun <A> genK(gen: Gen<A>): Gen<Kind<ForIO, A>> =
+    gen.map {
+      IO.just(it)
```

Comment:
We're only using 1 constructor here. Even if we're only interested in creating pure values shouldn't we also consider the other constructors? [Cats-effects](https://github.com/typelevel/cats-effect/blob/master/laws/shared/src/main/scala/cats/effect/laws/discipline/arbitrary.scala#L39) has quite an elaborate `Arbitrary` instance which also creates `IO` with depth.

```kotlin
internal fun IO.Companion.genK() = object : GenK<ForIO> {
  override fun <A> genK(gen: Gen<A>): Gen<Kind<ForIO, A>> = Gen.oneOf(
    gen.map(IO.Companion::just),
    gen.map { a -> IO.async<A> { cb -> cb(Right(a)) } },
    gen.map { a -> IO.effect { a } },
    gen.map { a -> IO.later { a } },
    gen.map { a -> IO.cancelable<A> { cb -> cb(Right(a)); IO.unit } },
    Gen.throwable().map(IO.Companion::raiseError)
  )
}
```

Reply:
Thats a good idea ^^.

Btw if we really want to we can go quite a bit further on this later on with arrow-check. I am talking actual side effects (like having `SideEffect` generate inside `StateT` for example). `GenT<StateTPartialOf<ForIO, List<SideEffect>>, A>`, with the machinery in `arrow-check` I can easily turn this into `StateT<GenTPartialOf<ForIO>, List<SideEffect>, A>` and back. This means two things: There is an instance for `MonadGen` for `StateT` and I can access and use the without much overhead directly in the test (no mess of logic needed to evaluate state all throughout^^).
OR you could go the other way and generate side-effects along `IO` programs that affect it and use `StateT<PropertyTPartialOf<ForIO>, List<SideEffect>, A>` (hidden behind `StateT.monadTest()` so you never see those ugly types. Same with the `GenT` part, it will always be used throgh `monadGen()` anyway). This means we can build up a state throughout testing and keep modifying it with our generated programs. This method is much more useful when doing stateful property tests rather than this `IO` example, but it shows just how powerful mtl can be. And hidden behind the typeclasses it will be invisible to anyone who doesn't care about it^^